src/test/scala/uk/gov/nationalarchives/dp/client/ClientTest.scala

```diff
@@ -0,0 +1,460 @@
+package uk.gov.nationalarchives.dp.client
+
+import cats.MonadError
+import com.github.tomakehurst.wiremock.WireMockServer
+import com.github.tomakehurst.wiremock.client.WireMock._
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers._
+import org.scalatest.{Assertion, BeforeAndAfterEach}
+import sttp.capabilities.Streams
+import uk.gov.nationalarchives.dp.client.Client.AuthDetails
+import uk.gov.nationalarchives.dp.client.Entity.fromType
+
+import java.time.{ZoneId, ZonedDateTime}
+import java.util.UUID
+import scala.jdk.CollectionConverters._
+
+abstract class ClientTest[F[_], S](port: Int, stream: Streams[S])(implicit
+    cme: MonadError[F, Throwable]
+) extends AnyFlatSpec
+    with BeforeAndAfterEach {
+
+  def valueFromF[T](value: F[T]): T
+
+  def createClient(url: String): F[Client[F, S]]
+
+  def testClient(url: String): Client[F, S] = valueFromF(createClient(url))
+
+  override def beforeEach(): Unit = {
+    preservicaServer.start()
+    preservicaServer.resetAll()
+  }
+
+  override def afterEach(): Unit = preservicaServer.stop()
+
+  val preservicaServer = new WireMockServer(port)
+
+  def checkServerCall(url: String): Assertion =
+    preservicaServer.getAllServeEvents.asScala.count(_.getRequest.getUrl == url) should equal(1)
+
+  val authDetails: AuthDetails = AuthDetails("user", "password")
+  val tokenResponse: String = """{"token": "abcde"}"""
+  val tokenUrl = "/api/accesstoken/login"
+
+  "getBitstreamInfo" should "call the correct API endpoints and return the bitstream info" in {
+    val id = UUID.randomUUID()
+    val fileName = "test.txt"
+
+    val entityUrl = s"/api/entity/content-objects/$id"
```

Comment:
1. A few of these variables repeat, do you think it might be better to make them properties of the class?
2. You are using `id`, should we use `ref` instead?

Reply:
I've moved `entityUrl` and `id` to the top of the class. There are other `entityUrl`s but they're using other values so I've left them there.